lib/pacemaker/pcmk_sched_location.c

```diff
@@ -289,38 +290,8 @@ unpack_rsc_location(xmlNode *xml_obj, pcmk_resource_t *rsc,
             .rsc_id_nmatches = rsc_id_nmatches,
         };
 
-        /* This loop is logically parallel to pcmk__evaluate_rules(), except
-         * instead of checking whether any rule is active, we set up location
-         * constraints for each active rule.
-         *
-         * @COMPAT When we can break backward compatibility, limit location
-         * constraints to a single rule, for consistency with other contexts.
-         * Since a rule may contain other rules, this does not prohibit any
-         * existing use cases.
-         *
-         * The schema already limits location constraints to a single rule.
-         */
-        for (xmlNode *rule_xml = pcmk__xe_first_child(xml_obj, PCMK_XE_RULE,
-                                                      NULL, NULL);
-             rule_xml != NULL; rule_xml = pcmk__xe_next_same(rule_xml)) {
-
-            if (generate_location_rule(rsc, rule_xml, discovery, next_change,
-                                       &rule_input)) {
-                if (empty) {
-                    empty = false;
-                    continue;
-                }
-                pcmk__warn_once(pcmk__wo_location_rules,
-                                "Support for multiple " PCMK_XE_RULE
-                                " elements in a location constraint is "
-                                "deprecated and will be removed in a future "
-                                "release (use a single new rule combining the "
-                                "previous rules with " PCMK_XA_BOOLEAN_OP
-                                " set to '" PCMK_VALUE_OR "' instead)");
-            }
-        }
-
-        if (empty) {
+        if (!generate_location_rule(rsc, rule_xml, discovery, next_change,
+                                    &rule_input)) {
             pcmk__config_err("Ignoring constraint '%s' because it contains "
                              "no valid rules", id);
```

Comment:
We can drop this message and just improve the messages in `generate_location_rule()`

Reply:
Do they even need improvement? Every `return false` case except memory allocation failure already has its own message.

Reply:
NVM, I see

Reply:
honestly seems easier to keep it in `unpack_rsc_location()` so that we can log the constraint ID without passing another argument

Reply:
```
    if (rule_xml == NULL) {
        pcmk__config_err("Ignoring location constraint '%s' without "
                         PCMK_XE_RULE " or " PCMK_XA_NODE " and " PCMK_XA_SCORE,
                         constraint_id);
        return false; 
    }
```
or just
```
    if (rule_xml == NULL) {
        pcmk__config_err("Ignoring location constraint '%s' without "
                         PCMK_XE_RULE,
                         constraint_id);
        return false; 
    }
```
(`generate_location_rule()` isn't aware of node/score but it's static and called once)